### ISSUE TITLE
i3bar freeze on displayport monitor enable

### ISSUE THREAD

**Issue by kb100** [i3/i3#1839]
The following consistently permanently freezes my i3bar, requiring 'i3-msg restart' to unfreeze:

    xrandr --output DP1 --off
    xrandr --output DP1 --auto --primary --output HDMI3 --auto --right-of DP1 --output VGA1 --auto --left-of DP1

Surprisingly this only happens for DP1, and not HDMI3 or VGA1.

Furthermore, the bar freezes when turning the output off, but always unfreezes after the monitor fully shuts off. However, when turning back on it freezes and never unfreezes.

Edit: It also occurs when my DP1 monitor turns itself off due to inactivity. This is how I noticed the bug, and it only started doing this in the past month (I want to say since 4.10.2).

i3 version 4.10.3 (2015-07-30, branch "4.10.3") © 2009-2014 Michael Stapelberg and contributors
http://logs.i3wm.org/logs/5639574185312256.bz2

**Comment by stapelberg** [comment]
There are only a handful of commits between the 4.10.x releases (see https://github.com/i3/i3/compare/4.10.1...4.10.3), so could you downgrade to 4.10.1 to see whether that fixes the issue? If downgrading helps, it’d be tremendously useful if you could use `git bisect` to narrow the problem down to a specific commit.

In case that’s not feasible or doesn’t help, can you please follow section 7 of http://build.i3wm.org/docs/debugging.html (“debugging i3bar”)?

**Comment by kb100** [comment]
Never built a project this big before and I don't really understand the build instructions on http://i3wm.org/docs/buildbot.html. Is there a simple make && make install command that would let me compile the version(s) I need? And is it safe to do so without interfering with my debian-packaged i3?

**Comment by kb100** [comment]
In the meantime, I follow the debugging i3bar steps through an entire freeze cycle and this was the only thing in the log after:

    [libi3] libi3/font.c Using Pango font DejaVu Sans Mono, size 10
    [libi3] libi3/font.c X11 root window dictates 96.000000 DPI

**Comment by stapelberg** [comment]
> Is there a simple make && make install command that would let me compile the version(s) I need?

Yes, the command is literally `make && make install`.

> And is it safe to do so without interfering with my debian-packaged i3?

`make install` will overwrite the i3 binary from Debian, but you can always use `apt-get install --reinstall i3-wm` to get it back.

**Comment by kb100** [comment]
It appears the issue goes back further than expected, I was not able to compile a version that did not have the issue.

    # bad: [a4ffb23d207ce4cf29fa3efe5e64e6ea2cc54a4c] Merge branch 'release-4.10.3'
    git bisect bad a4ffb23d207ce4cf29fa3efe5e64e6ea2cc54a4c
    # bad: [2dea50d469acd03e93b8231ec556e2012510d872] release i3 4.10.1
    git bisect bad 2dea50d469acd03e93b8231ec556e2012510d872
    # bad: [0b145d214b88cc2967f2180d9da1d71ec8b1646c] release i3 4.9.1
    git bisect bad 0b145d214b88cc2967f2180d9da1d71ec8b1646c

I'm not able to compile versions earlier than that, when I try to make I get

    [i3] CC src/commands_parser.c
    /usr/bin/ld: cannot find -lXcursor
    collect2: error: ld returned 1 exit status
    src/i3.mk:43: recipe for target 'src/commands_parser.o' failed
    make: *** [src/commands_parser.o] Error 1

**Comment by stapelberg** [comment]
You can `apt-get install libxcursor-dev` to make that library available.

**Comment by kb100** [comment]
Now I get the following errors on compiling (tried multiple versions):

    [i3] CC src/cfgparse.tab.c
    src/cfgparse.y: In function ‘start_configerror_nagbar’:
    src/cfgparse.y:262:35: error: ‘TERM_EMU’ undeclared (first use in this function)
    src/cfgparse.y:262:35: note: each undeclared identifier is reported only once for each function it appears in
    src/cfgparse.y:262:44: error: expected ‘)’ before string constant
    src/cfgparse.y:264:45: error: expected ‘)’ before string constant
    src/cfgparse.y: In function ‘yyparse’:
    src/cfgparse.y:691:29: warning: assignment from incompatible pointer type
    src/cfgparse.y:726:28: warning: assignment from incompatible pointer type
    src/cfgparse.y:731:29: warning: assignment from incompatible pointer type
    src/cfgparse.y:871:24: error: ‘BS_1PIXEL’ undeclared (first use in this function)
    src/cfgparse.y:977:26: warning: assignment from incompatible pointer type
    src/cfgparse.y:980:26: warning: assignment from incompatible pointer type
    src/cfgparse.y:983:9: warning: format ‘%s’ expects argument of type ‘char *’, but argument 2 has type ‘struct regex *’ [-Wformat=]
    src/cfgparse.y:984:9: warning: format ‘%s’ expects argument of type ‘char *’, but argument 2 has type ‘struct regex *’ [-Wformat=]
    src/i3.mk:36: recipe for target 'src/cfgparse.tab.o' failed
    make: *** [src/cfgparse.tab.o] Error 1

**Comment by stapelberg** [comment]
The version you’re trying seems way too old. We haven’t used a lex/yacc-based parser for quite a while. Try removing src/*.tab.c, but don’t bother with such an old version if that doesn’t help.

**Comment by kb100** [comment]
I figured out that I needed to be make cleaning inbetween checkouts (my bad for not knowing how git works). In process of bisecting now.

**Comment by kb100** [comment]
The problem seems more intermittent than previously thought as well. Here's my bisection log, I think I've narrowed it down to the exact commit. Note: "good" means I tried for a long time and didn't get a freeze. "bad" means there was a time when I got a freeze. Thus all the "bads" are definitely correct, whereas a "good" might be wrong.

    git bisect start
    # good: [30992e48d82ef6a3d244d93d720247388bee7c05] man: bump version to 4.8
    git bisect good 30992e48d82ef6a3d244d93d720247388bee7c05
    # bad: [55ba1c5e56ccddc901a8afd653fa1710035a2884] release i3 v4.9
    git bisect bad 55ba1c5e56ccddc901a8afd653fa1710035a2884
    # bad: [cc09348414472e68287ab918df57158552146f72] Merge branch 'master' into next
    git bisect bad cc09348414472e68287ab918df57158552146f72
    # good: [01d8d10c57cd09c73ce45e53f5d68cdbb7de268d] libxkbcommon 0.4.0 is required, 0.2.0 is too old (Thanks Mii)
    git bisect good 01d8d10c57cd09c73ce45e53f5d68cdbb7de268d
    # good: [196e748e94550b0f31714517526147a47f922ee5] bugfix: default floating border regression
    git bisect good 196e748e94550b0f31714517526147a47f922ee5
    # bad: [1842ddac51b9f4c79cb2d97a953e2a186012fd1a] Merge branch 'fix-tray-restart'
    git bisect bad 1842ddac51b9f4c79cb2d97a953e2a186012fd1a
    # good: [679a5de8cfb9211afb1a1dd882e53bd672084754] Bugfix: properly restore workspace containers (Thanks vals)
    git bisect good 679a5de8cfb9211afb1a1dd882e53bd672084754
    # bad: [a62bb17511949dbc6880763360dd2a530ec70866] Merge branch 'fix-decoration'
    git bisect bad a62bb17511949dbc6880763360dd2a530ec70866
    # bad: [56c284af8e281d9d5657e55b3edb7f7c673fc747] Bugfix: missed case for decoration rerender
    git bisect bad 56c284af8e281d9d5657e55b3edb7f7c673fc747
    # good: [9463be42b84b2d87b470e0ef5ccf3d24114e791a] Merge branch 'fix-layout-output'
    git bisect good 9463be42b84b2d87b470e0ef5ccf3d24114e791a
    # first bad commit: [56c284af8e281d9d5657e55b3edb7f7c673fc747] Bugfix: missed case for decoration rerender

**Comment by kb100** [comment]
[30992e48d82ef6a3d244d93d720247388bee7c05]
[55ba1c5e56ccddc901a8afd653fa1710035a2884]
[cc09348414472e68287ab918df57158552146f72]
[01d8d10c57cd09c73ce45e53f5d68cdbb7de268d]
[196e748e94550b0f31714517526147a47f922ee5]
[1842ddac51b9f4c79cb2d97a953e2a186012fd1a]
[679a5de8cfb9211afb1a1dd882e53bd672084754]
[a62bb17511949dbc6880763360dd2a530ec70866]
[56c284af8e281d9d5657e55b3edb7f7c673fc747]
[9463be42b84b2d87b470e0ef5ccf3d24114e791a]
[56c284af8e281d9d5657e55b3edb7f7c673fc747]

The commit hashes so github can link them for your

**Comment by kb100** [comment]
Another noticed symptom is that when a crash happens, only two of the three bars freeze/crash. The bar on the display port monitor doesn't freeze. This also happens to be the monitor that powers off first/quickest. I noticed because when my monitor goes to sleep after idling, all the workspaces on this monitor get moved to another monitor.

**Comment by stapelberg** [comment]
The code which is changed in commit 56c284af8e281d9d5657e55b3edb7f7c673fc747 is not even compiled into i3bar at all, so I don’t think the bisection results are reliable.

When i3bar is frozen, does it consume 100% CPU in `top(1)`?

When i3bar is frozen, can you attached gdb (`gdb =i3bar <pid>`) and provide the output of `backtrace full`?

**Comment by kb100** [comment]
When it is frozen, it only freezes on 2/3 of my monitors, it continues working fine on the monitor that I'm turning on and off.

As for CPU, it says i3bar is using 0 CPU.

Backtrace:

    (gdb) backtrace full
    #0  0x00007f9e9a414633 in __epoll_wait_nocancel () at ../sysdeps/unix/syscall-template.S:81
    No locals.
    #1  0x00007f9e9a8f7dd5 in ?? () from /usr/lib/x86_64-linux-gnu/libev.so.4
    No symbol table info available.
    #2  0x00007f9e9a8fa068 in ev_run () from /usr/lib/x86_64-linux-gnu/libev.so.4
    No symbol table info available.
    #3  0x0000000000403ce7 in ev_loop (flags=0, loop=<optimized out>) at /usr/include/ev.h:826
    No locals.
    #4  main (argc=<optimized out>, argv=<optimized out>) at i3bar/src/main.c:170
            opt = <optimized out>
            option_index = 0
            socket_path = <optimized out>
            i3_default_sock_path = 0x40f09f "/tmp/i3-ipc.sock"
            long_opt = {{name = 0x410911 "socket", has_arg = 1, flag = 0x0, val = 115}, {name = 0x40f0b0 "bar_id", has_arg = 1, flag = 0x0, val = 98}, {name = 0x40f0b7 "help", has_arg = 0, flag = 0x0, val = 104}, {name = 0x40e43c "version", has_arg = 0, flag = 0x0, val = 118}, {
                name = 0x0, has_arg = 0, flag = 0x0, val = 0}}
            atom_sock_path = <optimized out>
            sig_term = 0x12410f0
            sig_int = 0x1241130
            sig_hup = 0x1241170

**Comment by kb100** [comment]
It's extra strange because two of my monitors are identical, literally the same product bought in the same store at the same time, but I can only seem to trigger the bug turning on/off one of them. The only difference configuration-wise is that one is hooked up by displayport, the other is hdmi, and one is set as the "primary" monitor, and the other is not.

**Comment by stapelberg** [comment]
One difference might be that the monitor is detected as disconnected (or not) depending on how it’s connected, see http://permalink.gmane.org/gmane.comp.window-managers.i3.general/1834

**Comment by kb100** [comment]
I'm also suffering from that problem too, when my monitor turns off (still plugged in), all the workspaces from the DP monitor get moved. Could this possibly be related? I first noticed the bug in 4.10.3...

**Comment by kb100** [comment]
I used the following to cause a freeze (dpoff/dponn just call xrandr and turn the monitor on and off)
 
   i3-dump-log > before.log && sleep 3 && dpoff && sleep 35 && i3-dump-log > mid.log && sleep 5 && dponn && sleep 10 && i3-dump-log > after.log

Here is a diff of after.log and mid.log
http://logs.i3wm.org/logs/5649521866440704.bz2

and of mid.log and before.log
http://logs.i3wm.org/logs/5733935958982656.bz2

The after-mid diff looks much more interesting. In particular I notice that the log spans more than one second, possibly explaining why the bug only happens some of the time -- I'm guessing that a bar script fires on the second and tries to output before the monitor's status is properly set. 

Do these  logs give you more insight?

**Comment by TatriX** [comment]
My i3 completely freezes when I turn off and then turn on a monitor. I can use latest focused window and some hotkeys. i3-dump-log says there is no log file in that state. 4.10.1 has no such problem.
(archlinux x64,  i3-git 4.10.2.r223.gb30d87e-1)

**Comment by kb100** [comment]
@TatriX see http://build.i3wm.org/docs/debugging.html for how to enable the debug log. Also, in my case the problem is intermittent so try turning it off and back on again many times, waiting for varying amounts of time inbetween. I was able to produce the bug going way back. See if you can reproduce.

**Comment by TatriX** [comment]
When debugging enabled and freeze occurs i3-dump-log says that there is no log found. If you need debug log from non bugged session I can make one.

**Comment by stapelberg** [comment]
I haven’t had a chance to look at the logs yet (will do later), but:

> When debugging enabled and freeze occurs i3-dump-log says that there is no log found.

Can you nopaste the specific error message it provides please?

Also, what happens when you use `i3-dump-log -f > /tmp/i3.log` before triggering the freeze/crash?

If you enable core dumps (see section 3 of http://i3wm.org/docs/debugging-release-version.html), do you get a core dump file? If yes, can you attach the backtrace please (see section 5 of http://i3wm.org/docs/debugging-release-version.html)?

**Comment by TatriX** [comment]
```i3-dump-log -f > /tmp/i3.log``` helped:
http://logs.i3wm.org/logs/5689154616688640.bz2
``` PM - sighandler.c:handle_signal:292 - i3 crashed. SIG: 11```
<pre>i3 version 4.10.3 (2015-07-30, branch "4.10.3") © 2009-2014 Michael Stapelberg and contributors
[1]    11754 segmentation fault (core dumped)  i3 -v
</pre>

**Comment by stapelberg** [comment]
@TatriX Okay. Can you open a separate issue with the backtrace for this core dump please? While it might be related, I don’t think it’s the same issue as #1839 (this one).

**Comment by TatriX** [comment]
Yeah, sure. Also latest version have my hotkeys broken, so I'm going to open two issues.

**Comment by stapelberg** [comment]
@kb100, I think we should focus on getting the i3bar log to work, so that we can see how i3bar reacts to the disappearing output.

Also, can you please provide a screenshot of when i3bar is frozen to better illustrate what you mean by frozen? Is it still shown, it just doesn’t update?

> In the meantime, I follow the debugging i3bar steps through an entire freeze cycle and this was the only thing in the log after:
> 
> [libi3] libi3/font.c Using Pango font DejaVu Sans Mono, size 10
> [libi3] libi3/font.c X11 root window dictates 96.000000 DPI

This looks like you didn’t add the `verbose yes` to your bar block. If you did (and reloaded i3 afterwards, that might not be clear from the debug guide), can you also provide the i3 debug log please?

**Comment by kb100** [comment]
Here is a paste of the i3bar log, your log server wouldn't accept it so I just used pastebin. I removed all the "full_text", "short_text", and "color = " lines as pastebin wouldn't let me post such a huge file.

http://pastebin.com/download.php?i=iX0NVq65

One thing that immediately pops out to me is that turning the monitor on, which I can track by eye in the file by looking for "reconfiguring windows", actually triggers three reconfigure events in rapid succession. (on/off/on). I wonder if they are actually running in parallel and interfering with each other.

Here's a picture of a freeze:

![freeze](https://cloud.githubusercontent.com/assets/1966710/9570544/3f66b4fc-4f54-11e5-8f23-20ad1ca696af.png)

The leftmost bar is truly frozen, it looks normal except it is not being updated and does not respond to clicks. The middle bar is not frozen, it corresponds to the displayport monitor and is fully functional, clicks and all.. The rightmost bar is frozen and scrambled. It appears to be showing chunks of Deluge, which I had on a workspace on that monitor. Deluge was NOT on the shown workspace on that monitor at the time of scrambling.

**Comment by kb100** [comment]
I haven't changed my version of i3 (still 4.10.3) but now the non DP monitors don't seem to freeze anymore. Instead, the same script that used to cause a freeze now just moves all the workspaces from the DP monitor to the next monitor. Since I haven't upgraded my i3, I now suspect that the bug may arise from an interplay between i3 and another package (probably an X related package). I will keep a close watch for when

https://github.com/i3/i3/commit/1a9a9cc68d2eae4058983e18ca3cbc52e9d1abb1

makes it into debian stretch to see if the bug was really not an i3 bug at all.

**Comment by stapelberg** [comment]
@kb100 i3 4.10.4 is now in Debian stretch. Can you update us on how that changes things?

**Comment by kb100** [comment]
I have not had a freeze since 4.10.4, but I haven't done extensive testing. I'll check back in a week or so and let you know.

**Comment by Airblader** [comment]
@kb100 Now it's been a week. What's the status? :)

**Comment by kb100** [comment]
:( freeze on first attempt using my script from above. It hasn't occurred naturally though since https://github.com/i3/i3/commit/ce7d256d05adea3d17eb16b336a3237fb5529ea2. Meaning I used my computer for a week with no freeze. Before the revert I was getting freezes every day. I went a full week with no freeze. But, if I turn the monitor off and back on with the script I gave at the top, I can still trigger a freeze.

**Comment by kb100** [comment]
As I discovered in my git bisection, the bug goes very far back, so it wasn't introduced in the now-reverted commit, though that commit may have exacerbated the problem.

**Comment by kb100** [comment]
@stapelberg I haven't experienced the bug in months now so I'm closing this issue. Will you please delete the logs

http://logs.i3wm.org/logs/5639574185312256.bz2
http://logs.i3wm.org/logs/5649521866440704.bz2
http://logs.i3wm.org/logs/5733935958982656.bz2

which contain personally identifiable information.

**Comment by stapelberg** [comment]
Done.

**Comment by iaculch** [comment]
I am still experiencing this freeze, it is quite intermittent, happening to me a couple of times a week, and I am plugging in an external monitor to my laptop a few times a day.

This is the command I run to activate the external monitor:

`xrandr --output eDP1 --mode 3200x1800 --pos 288x2160 --rotate normal --scale 0.5x0.5 --output DP1 --mode 1920x1080 --pos 0x0 --rotate normal`

and another to switch back to solely my laptop display:

`xrandr --output eDP1 --mode 3200x1800 --pos 288x2160 --rotate normal --scale 0.5x0.5 --output DP1 --off`
